packages/headless/src/controllers/triggers/headless-query-trigger.ts

```diff
@@ -0,0 +1,93 @@
+import {SearchEngine} from '../../app/search-engine/search-engine';
+import {TriggerSection} from '../../state/state-sections';
+import {triggers} from '../../app/reducers';
+import {buildController, Controller} from '../controller/headless-controller';
+import {loadReducerError} from '../../utils/errors';
+import {executeSearch} from '../../features/search/search-actions';
+import {
+  updateQuery,
+  UpdateQueryActionCreatorPayload,
+} from '../../features/query/query-actions';
+import {logQueryTrigger} from '../../features/triggers/trigger-analytics-actions';
+
+/**
+ * The `QueryTrigger` controller handles query triggers.
+ */
+export interface QueryTrigger extends Controller {
+  /**
+   * the state of the `QueryTrigger` controller.
+   */
+  state: QueryTriggerState;
+}
+
+/**
+ * A scoped and simplified part of the headless state that is relevant to the `QueryTrigger` controller.
+ */
+export interface QueryTriggerState {
+  /**
+   * The new query to perform a search with after receiving a query trigger.
+   */
+  newQuery: string;
+
+  /**
+   * The query used to perform the search that received a query trigger in its response.
+   */
+  prevQuery: string | undefined;
+}
+
+/**
+ * Creates a `QueryTrigger` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @returns A `QueryTrigger` controller instance.
+ * */
+export function buildQueryTrigger(engine: SearchEngine): QueryTrigger {
+  if (!loadQueryTriggerReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+
+  const getState = () => engine.state;
+
+  let previousQueryTrigger: string = getState().triggers.query;
+  let previousQuery: string | undefined = getState().query?.q;
+
+  return {
+    ...controller,
+
+    subscribe(listener: () => void) {
+      const strictListener = () => {
+        const hasChanged = previousQueryTrigger !== this.state.newQuery;
+        previousQueryTrigger = this.state.newQuery;
+        previousQuery = getState().query?.q;
```

Comment:
If you update this value all the time, you'll have the same original problem: next time an action is dispatched previousQuery = currentQuery

The value should be updated after the if

Reply:
Ah yes, I didn't think of actions not related to the triggers possibly being dispatched.  I guess I forgot this will run whenever there is a state change not only when there is a trigger state change. Perhaps, it would be useful to create subscribe methods that only watch a specific part of the engine state?

Reply:
That's the purpose of the controller subscribe method 😄 
We subscribe to the engine, and only call the listener when relevant parts of the state change.

Reply:
If it's updated after the if,  then it will take on the newly updated query value from the Trigger and we won't have access to the previous query.